File: services/media-service/internal/domain/domain.go
package domain

import (
	"context"
	"time"

	"github.com/google/uuid"
)

type EventProducer interface {
	Publish(ctx context.Context, eventType string, data interface{})
	Close()
}

type ContentType string

const (
	ContentTypeVideo ContentType = "video"
	ContentTypeImage ContentType = "image"
	ContentTypeAudio ContentType = "audio"
)

type MediaStatus string

const (
	MediaStatusUploading  MediaStatus = "uploading"
	MediaStatusProcessing MediaStatus = "processing"
	MediaStatusReady      MediaStatus = "ready"
	MediaStatusFailed     MediaStatus = "failed"
	MediaStatusDeleted    MediaStatus = "deleted"
)

type MediaAsset struct {
	ID             uuid.UUID   `json:"id"`
	UploaderID     uuid.UUID   `json:"uploaderId"`
	ContentType    ContentType `json:"contentType"`
	SourceFilename string      `json:"sourceFilename"`
	FileSizeBytes  int64       `json:"fileSizeBytes"`
	MimeType       string      `json:"mimeType"`
	Status         MediaStatus `json:"status"`
	StoragePath    string      `json:"storagePath"`
	CDNURL         string      `json:"cdnUrl"`
	DurationSecs   float64     `json:"durationSeconds"`
	Width          int         `json:"width"`
	Height         int         `json:"height"`
	Checksum       string      `json:"checksum"`
	CreatedAt      time.Time   `json:"createdAt"`
}

type RenditionProfile string

const (
	RenditionThumbnail RenditionProfile = "thumbnail"
	Rendition720p      RenditionProfile = "720p"
	Rendition1080p     RenditionProfile = "1080p"
	RenditionSource    RenditionProfile = "source"
)

type RenditionFormat string

const (
	RenditionHLS   RenditionFormat = "hls"
	RenditionDASH  RenditionFormat = "dash"
	RenditionMP4   RenditionFormat = "mp4"
	RenditionWebP  RenditionFormat = "webp"
	RenditionJPG   RenditionFormat = "jpg"
)

type MediaRendition struct {
	ID           uuid.UUID        `json:"id"`
	MediaID      uuid.UUID        `json:"mediaId"`
	Profile      RenditionProfile `json:"profile"`
	Format       RenditionFormat  `json:"format"`
	FileSizeBytes int64           `json:"fileSizeBytes"`
	StoragePath  string           `json:"storagePath"`
	CDNURL       string           `json:"cdnUrl"`
	Status       MediaStatus      `json:"status"`
	CreatedAt    time.Time        `json:"createdAt"`
}

type TranscodingJobStatus string

const (
	TranscodingPending    TranscodingJobStatus = "pending"
	TranscodingProcessing TranscodingJobStatus = "processing"
	TranscodingCompleted  TranscodingJobStatus = "completed"
	TranscodingFailed     TranscodingJobStatus = "failed"
)

type TranscodingJob struct {
	ID             uuid.UUID            `json:"id"`
	MediaID        uuid.UUID            `json:"mediaId"`
	Profiles       []RenditionProfile   `json:"profiles"`
	Status         TranscodingJobStatus `json:"status"`
	WorkerID       string               `json:"workerId"`
	StartedAt      *time.Time           `json:"startedAt,omitempty"`
	CompletedAt    *time.Time           `json:"completedAt,omitempty"`
	ErrorMessage   string               `json:"errorMessage,omitempty"`
	CreatedAt      time.Time            `json:"createdAt"`
}

type KeySystem string

const (
	KeySystemWidevine KeySystem = "widevine"
	KeySystemFairPlay KeySystem = "fairplay"
)

type DRMPolicy struct {
	ID                   uuid.UUID  `json:"id"`
	Name                 string     `json:"name"`
	ContentID            *uuid.UUID `json:"contentId,omitempty"`
	KeySystem            KeySystem  `json:"keySystem"`
	LicenseDurationSecs  int64      `json:"licenseDurationSeconds"`
	SecurityLevel        string     `json:"securityLevel"`
	IsActive             bool       `json:"isActive"`
	CreatedAt            time.Time  `json:"createdAt"`
}

type UploadSession struct {
	ID            uuid.UUID `json:"id"`
	UploaderID    uuid.UUID `json:"uploaderId"`
	Filename      string    `json:"filename"`
	FileSizeBytes int64     `json:"fileSizeBytes"`
	ContentType   string    `json:"contentType"`
	ChunksTotal   int       `json:"chunksTotal"`
	ChunksDone    int       `json:"chunksDone"`
	Checksum      string    `json:"checksum"`
	Status        string    `json:"status"`
	StoragePath   string    `json:"storagePath"`
	CreatedAt     time.Time `json:"createdAt"`
	ExpiresAt     time.Time `json:"expiresAt"`
}

type StorageUsage struct {
	TotalBytes   int64 `json:"totalBytes"`
	UploadBytes  int64 `json:"uploadBytes"`
	RenditionBytes int64 `json:"renditionBytes"`
	ThumbnailBytes int64 `json:"thumbnailBytes"`
	AssetCount   int64 `json:"assetCount"`
}

File: services/media-service/src/domain.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "snake_case")]
pub enum ContentType {
    Video,
    Image,
    Audio,
}

impl std::fmt::Display for ContentType {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            ContentType::Video => write!(f, "video"),
            ContentType::Image => write!(f, "image"),
            ContentType::Audio => write!(f, "audio"),
        }
    }
}

impl From<&str> for ContentType {
    fn from(s: &str) -> Self {
        match s {
            "video" => ContentType::Video,
            "image" => ContentType::Image,
            "audio" => ContentType::Audio,
            _ => ContentType::Video,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "snake_case")]
pub enum MediaStatus {
    Uploading,
    Processing,
    Ready,
    Failed,
    Deleted,
}

impl std::fmt::Display for MediaStatus {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            MediaStatus::Uploading => write!(f, "uploading"),
            MediaStatus::Processing => write!(f, "processing"),
            MediaStatus::Ready => write!(f, "ready"),
            MediaStatus::Failed => write!(f, "failed"),
            MediaStatus::Deleted => write!(f, "deleted"),
        }
    }
}

impl From<&str> for MediaStatus {
    fn from(s: &str) -> Self {
        match s {
            "uploading" => MediaStatus::Uploading,
            "processing" => MediaStatus::Processing,
            "ready" => MediaStatus::Ready,
            "failed" => MediaStatus::Failed,
            "deleted" => MediaStatus::Deleted,
            _ => MediaStatus::Uploading,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MediaAsset {
    pub id: Uuid,
    pub uploader_id: Uuid,
    pub content_type: ContentType,
    pub source_filename: String,
    pub file_size_bytes: i64,
    pub mime_type: String,
    pub status: MediaStatus,
    pub storage_path: String,
    pub cdn_url: String,
    pub duration_seconds: f64,
    pub width: i32,
    pub height: i32,
    pub checksum: String,
    pub created_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MediaRendition {
    pub id: Uuid,
    pub media_id: Uuid,
    pub profile: String,
    pub format: String,
    pub file_size_bytes: i64,
    pub storage_path: String,
    pub cdn_url: String,
    pub status: MediaStatus,
    pub created_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "snake_case")]
pub enum TranscodingStatus {
    Pending,
    Processing,
    Completed,
    Failed,
}

impl std::fmt::Display for TranscodingStatus {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            TranscodingStatus::Pending => write!(f, "pending"),
            TranscodingStatus::Processing => write!(f, "processing"),
            TranscodingStatus::Completed => write!(f, "completed"),
            TranscodingStatus::Failed => write!(f, "failed"),
        }
    }
}

impl From<&str> for TranscodingStatus {
    fn from(s: &str) -> Self {
        match s {
            "pending" => TranscodingStatus::Pending,
            "processing" => TranscodingStatus::Processing,
            "completed" => TranscodingStatus::Completed,
            "failed" => TranscodingStatus::Failed,
            _ => TranscodingStatus::Pending,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TranscodingJob {
    pub id: Uuid,
    pub media_id: Uuid,
    pub profiles: Vec<String>,
    pub status: TranscodingStatus,
    pub worker_id: String,
    pub started_at: Option<DateTime<Utc>>,
    pub completed_at: Option<DateTime<Utc>>,
    pub error_message: String,
    pub created_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DrmPolicy {
    pub id: Uuid,
    pub name: String,
    pub content_id: Option<Uuid>,
    pub key_system: String,
    pub license_duration_seconds: i64,
    pub security_level: String,
    pub is_active: bool,
    pub created_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UploadSession {
    pub id: Uuid,
    pub uploader_id: Uuid,
    pub filename: String,
    pub file_size_bytes: i64,
    pub content_type: String,
    pub chunks_total: i32,
    pub chunks_done: i32,
    pub checksum: String,
    pub status: String,
    pub storage_path: String,
    pub created_at: DateTime<Utc>,
    pub expires_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct StorageUsage {
    pub total_bytes: i64,
    pub upload_bytes: i64,
    pub rendition_bytes: i64,
    pub thumbnail_bytes: i64,
    pub asset_count: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct InitUploadRequest {
    pub uploader_id: Uuid,
    pub filename: String,
    pub content_type: String,
    pub file_size_bytes: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UploadChunkRequest {
    pub chunk_index: i32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TranscodeRequest {
    pub media_id: Uuid,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ThumbnailRequest {
    pub media_id: Uuid,
    pub time_seconds: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OptimizeRequest {
    pub media_id: Uuid,
    pub width: i32,
    pub height: i32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DrmLicenseRequest {
    pub media_id: Uuid,
    pub key_system: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CachePurgeRequest {
    pub path: String,
}
